actionview/lib/action_view/template/renderable.rb

```diff
@@ -4,20 +4,22 @@ module ActionView
   class Template
     # = Action View Renderable Template for objects that respond to #render_in
     class Renderable # :nodoc:
-      def initialize(renderable)
+      def initialize(renderable, formats, &block)
         @renderable = renderable
+        @formats = formats
+        @block = block
       end
 
       def identifier
         @renderable.class.name
       end
 
-      def render(context, *args)
-        @renderable.render_in(context)
+      def render(context, locals)
+        @renderable.render_in(context, formats: @formats, locals: locals, &@block)
       end
 
       def format
-        @renderable.format
+        @renderable.try(:format) || Array(@formats).first
```

Comment:
Why are we trying to call the format now?

Reply:
@rafaelfranca this is related to https://github.com/rails/rails/pull/50622.

When `#render_in` was introduced, there wasn't any public documentation citing that in addition to `#render_in`, objects would also need to define `#format`.

The `TestRenderable` in the test suite defines it (and, for example, I think `ViewComponent::Base` also defines it). It's called internal to Action View, so it's part of the public API despite the fact that it wasn't part of any of the published documentation.

https://github.com/rails/rails/pull/50622/files#diff-70344644d049724d12c9a0c2c387dd1ea31663fbfad4c6a1acdd49efc3cb9512R20 is actually the source of the `Object#try` change. The motivation behind that change was that since it was newly documented, prior implementers of `#render_in` could get by without defining `#format`. The absence of the method means that `TemplateRenderable#format` returns `nil`, and somewhere down the chain the content gets treated as `:html` by default.

Since this change makes more of the options available to the `#render_in` call, it could be useful for the caller to decide how to render itself (as HTML, JSON, etc) based on that newly available information.

If the `Object#try` doesn't feel correct, would we be able to change this interface to accept the `:formats` available from the `render` call as arguments?

Reply:
@rafaelfranca based on [this discussion](https://github.com/rails/rails/pull/50622#discussion_r1445897182), I've made two changes:

* omit support for `render formats: ...` from this PR
* deprecation warning for `@renderable` objects that don't respond to `#format`

Reply:
> deprecation warning for @renderable objects that don't respond to #format
Why we need that?